crates/bevy_ecs/src/system/system_registry.rs

```diff
@@ -332,25 +337,37 @@ impl World {
         I: SystemInput + 'static,
         O: 'static,
     {
+        self.init_resource::<RunSystemStack>();
+
+        if self.resource::<RunSystemStack>().contains(&id.entity) {
+            return Err(RegisteredSystemError::Recursive(id));
+        }
+
         // Lookup
         let mut entity = self
             .get_entity_mut(id.entity)
             .map_err(|_| RegisteredSystemError::SystemIdNotRegistered(id))?;
 
         // Take ownership of system trait object
-        let RegisteredSystem {
+        let Some(RegisteredSystem {
             mut initialized,
             mut system,
-        } = entity
-            .take::<RegisteredSystem<I, O>>()
-            .ok_or(RegisteredSystemError::Recursive(id))?;
+        }) = entity.take::<RegisteredSystem<I, O>>()
+        else {
+            return Err(RegisteredSystemError::MaybeIncorrectType(id));
+        };
 
-        // Run the system
+        // Initialize the system
         if !initialized {
             system.initialize(self);
             initialized = true;
         }
 
+        let Some(mut sys_stack) = self.get_resource_mut::<RunSystemStack>() else {
+            panic!("`RunSystemStack` should never be removed during the execution of a one-shot system.");
+        };
+        sys_stack.insert(id.entity);
```

Comment:
I think you can simplify the "or panic" logic with `expect` (here and below).  

```suggestion
        self.get_resource_mut::<RunSystemStack>()
            .expect("`RunSystemStack` should never be removed during the execution of a one-shot system.")
            .insert(id.entity);
```